Print git info

KangarooKoala opened the pull request:
Adds a gradle task which writes the latest commit, changed files, and untracked files to `git-info.txt` in the deploy directory and adds code to print the contents of that file.

KangarooKoala committed:
Print git info

KangarooKoala commented:
An example output file:
```
### Latest commit (git log -1 --decorate=short --format=medium)
commit 63d01454b95d4a098def03dd124ce9ea1ddac7b3 (HEAD -> robotInfo, origin/robotInfo)
Author: Joseph Eng <91924258+KangarooKoala@users.noreply.github.com>
Date:   Sun Apr 2 21:14:05 2023 -0700

    Print git info
### Changed files (git diff-index --name-only HEAD --)
src/main/java/frc/team2412/robot/Robot.java
### Untracked files (git ls-files --exclude-standard --others)
### (END)
```

TAKBS2412 reviewed (commented):
How does this handle setups where `git` isn't in the PATH? (Or are our school laptops and other computers set up to where that wouldn't be an issue?)

KangarooKoala commented:
This currently does not handle a lack of git, though I don't know under what circumstances we would be deploying or simulating robot code on a computer that does not have git.

bobtseattle reviewed (commented):
This is a great start, but seems like we could use more data in the output to better understand origination.
The things that tend to cause is grief are the git commit, the branch, the remote repo, the path of the source the, and the laptop / host where the code was built and deployed.  Though, knowing the source path the remote repo doesn't seem necessary.
It would be nice to have a checksum of the changed files to know when builds differ, but that is probably over doing it and not immediately available on school laptops.

obj-obj commented:
> This currently does not handle a lack of git, though I don't know under what circumstances we would be deploying or simulating robot code on a computer that does not have git.

I'm pretty sure BSD doesn't put git in the PATH. I remember having to do some goofy stuff last year to get the git executable from github desktop in the PATH so people's vscodes could use git

KangarooKoala commented:
What would be the desired behavior when git is not in the path?

>  the laptop / host where the code was built and deployed

What would distinguish between different laptops?

> knowing the source path the remote repo doesn't seem necessary.

What are you saying doesn't seem necessary? The source path, the remote repo, or both?

bobtseattle commented:
What would be the desired behavior when git is not in the path?
- Seems like it should print something along the lines of "git not in path so git status is not available."  It should still print the current working directory and hostname.  
- Some git clients don't use the native git executables in the background but use jgit.

| the laptop / host where the code was built and deployed
What would distinguish between different laptops
- Each machine should have a hostname.  Use 'hostname' from the CLI, or getHostName() from Java.  

| knowing the source path the remote repo doesn't seem necessary.
What are you saying doesn't seem necessary? The source path, the remote repo, or both?
- Assuming the directories on laptops don't often change which git remote they are pointing to, once we have the directory and hostname, we can figure out which repository or fork on github the code in the directory is based on.  The git remote -v is best, but the hostname and directory should generally be sufficient.

KangarooKoala committed:
Report more info, warn if not using git
Reports host name and project path
Warns when building if it can't run git
Also print out remotes

KangarooKoala commented:
Example output:
```
### Host name (InetAddress.localHost.hostName)
52S500******
### Project path (projectDir)
C:\Users\<username>\Documents\GitHub\ChargedUp2023
### Latest commit (git log -1 --decorate=short --format=medium)
commit a7ef919a5c813e3a3d8b656f5c5b280aa9dcf5f9 (HEAD -> robotInfo, origin/robotInfo, foo)
Author: Joseph Eng <91924258+KangarooKoala@users.noreply.github.com>
Date:   Sun Apr 2 21:14:05 2023 -0700

    Print git info
### Remotes (git remote -v)
jonah   https://github.com/jbko6/ChargedUp2023.git/ (fetch)
jonah   https://github.com/jbko6/ChargedUp2023.git/ (push)
joseph  https://github.com/KangarooKoala/ChargedUp2023.git (fetch)
joseph  https://github.com/KangarooKoala/ChargedUp2023.git (push)
kirby   https://github.com/kirbt/ChargedUp2023.git (fetch)
kirby   https://github.com/kirbt/ChargedUp2023.git (push)
origin  https://github.com/robototes/ChargedUp2023.git (fetch)
origin  https://github.com/robototes/ChargedUp2023.git (push)
### Changed files (git diff-index --name-only HEAD --)
build.gradle
### Untracked files (git ls-files --exclude-standard --others)
### (END)
```

Without git on PATH:
```
### Host name (InetAddress.localHost.hostName)
52S500******
### Project path (projectDir)
C:\Users\<username>\Documents\GitHub\ChargedUp2023
### <could not run git>
### (END)
```

Output when building without git:
![image](https://github.com/robototes/ChargedUp2023/assets/91924258/6e6ee3b7-8f7b-42c4-a767-0346027049fd)

KangarooKoala committed:
Split apart commit information
Name is omitted for privacy

KangarooKoala committed:
Replace username in project path

KangarooKoala committed:
Change path to forward slashes
This works on Windows and other platforms

TAKBS2412 commented:
(Reminder to myself to review and test this at a later time).

TAKBS2412 commented on build.gradle:
```diff
@@ -134,6 +136,72 @@ spotless {
+	}
+
+	def convertPath = { String path ->
+		String username = System.getenv("USERNAME")
```
(Minor as most of us use Windows) Can we switch this to `System.getProperty("user.name")` so we can support OSes that don't set the `USERNAME` environment variable? (I got `null` when testing in Ubuntu).

[Reference](https://stackoverflow.com/questions/38686481/how-to-get-userid-and-gid-in-gradle) (Ironically, I got the information needed from the question, and not the answer).

KangarooKoala replied:
I decided to leave both in so that if possible we could get something that could be pasted into a shell (assuming it's the same user).

KangarooKoala committed:
Use property user.name instead of environment variable USERNAME

KangarooKoala committed:
First check environment variable, then try user.name property
Also use different format for user.name property

bobtseattle reviewed (approved):
LGTM
